### PR TITLE
Support for LevelZero GPU CHIP temperatures

### LINKED ISSUE #2741: As a GPU research I would like Temperature controls via LevelZero
**What problem does the story solve? Please describe.**
I'm always frustrated when I want to read temperature data via LevelZero for Intel GPUs and cannot.  

**Describe the tasks that will be required to complete the story**
Provide access to GPU Chip level sensors as described [here](https://spec.oneapi.io/level-zero/latest/sysman/PROG.html#querying-temperature)

**Describe alternatives you've considered**
Using other interfaces than LevelZero or not providing temperature.

### PR DESCRIPTION
**Pull request by lhlawson**
- Relates to #2742 story from github issues
- Fixes #2741 change request from github issues.

Update to levelzero code to provide GPU_CHIP temperature information.

### CODE REVIEW

File: service/src/LevelZero.hpp [hunk 1]
```diff
@@ -76,6 +76,7 @@ namespace geopm
             /// @return GPU maximum frequency in MHz.
             virtual double frequency_max(unsigned int l0_device_idx, int l0_domain,
                                          int l0_domain_idx) const = 0;
+
```
Comment (bgeltz) [inline service/src/LevelZero.hpp]:
```suggestion
```

File: service/src/LevelZero.hpp [hunk 2]
```diff
@@ -97,6 +98,22 @@ namespace geopm
...
+            ///        Level Zero GPU.
+            /// @param [in] l0_domain The LevelZero domain type being targeted
+            /// @param [in] l0_domain_idx The LevelZero index indicating a particular
+            ///        domain of the GPU.
+            /// @return Domain maximum temperature in Celsius.
+            virtual double temperature_max(unsigned int l0_device_idx, int l0_domain,
+                                           int l0_domain_idx) const = 0;
+
+
+
```
Comment (bgeltz) [inline service/src/LevelZero.hpp]:
```suggestion
```

File: service/src/LevelZero.hpp [hunk 3]
```diff
@@ -97,6 +98,22 @@ namespace geopm
                                                               int l0_domain,
                                                               int l0_domain_idx) const = 0;
 
+            /// @brief Get the number of LevelZero temperature domains
+            /// @param [in] l0_domain The LevelZero domain type being targeted
+            /// @return GPU temperature domain count.
+            virtual int temperature_domain_count(unsigned int l0_device_idx, int l0_domain) const = 0;
```
Comment (bgeltz) [inline service/src/LevelZero.hpp]:
```suggestion
            /// @brief Get the number of LevelZero temperature domains
            /// @param [in] l0_device_idx The index indicating a particular
            ///        Level Zero GPU.
            /// @param [in] l0_domain The LevelZero domain type being targeted
            /// @return GPU temperature domain count.
            virtual int temperature_domain_count(unsigned int l0_device_idx, int l0_domain) const = 0;
```

File: service/src/LevelZeroDevicePool.hpp [hunk 4]
```diff
@@ -58,6 +58,7 @@ namespace geopm
             /// @return GPU maximum frequency in MHz.
             virtual double frequency_max(int domain, unsigned int domain_idx,
                                          int l0_domain) const = 0;
+
```
Comment (bgeltz) [inline service/src/LevelZeroDevicePool.hpp]:
```suggestion
```

File: service/src/LevelZeroIOGroup.cpp [hunk 5]
```diff
@@ -447,6 +447,38 @@ namespace geopm
                                   },
                                   .01
                                   }},
+                              {M_NAME_PREFIX + "GPU_CORE_TEMPERATURE_MAXIMUM", {
+                                  "The maximum measured temperature across all sensors in the GPU accelerator.",
+                                  GEOPM_DOMAIN_GPU_CHIP,
+                                  Agg::average,
```
Comment (bgeltz) [inline service/src/LevelZeroIOGroup.cpp]:
```suggestion
                                  Agg::max,
```
Does this matter in this case?
Perhaps if you issued ```geopmread GPU_CORE_TEMPERATURE_MAXIMUM board 0``` and you wanted to see the absolute maximum across all GPUs?  As it is, reading at the board domain would take the average of all the GPUs max and return that.

File: service/src/LevelZeroIOGroup.cpp [hunk 6]
```diff
@@ -447,6 +447,38 @@ namespace geopm
...
+                                                   GEOPM_DOMAIN_GPU_CHIP,
+                                                   domain_idx,
+                                                   geopm::LevelZero::M_DOMAIN_COMPUTE);
+                                  },
+                                  1
+                                  }},
+                              {M_NAME_PREFIX + "GPU_MEMORY_TEMPERATURE_MAXIMUM", {
+                                  "The maximum measured temperature across all sensors in the GPU memory.",
+                                  GEOPM_DOMAIN_GPU_CHIP,
+                                  Agg::average,
```
Comment (bgeltz) [inline service/src/LevelZeroIOGroup.cpp]:
```suggestion
                                  Agg::max,
```
Same question here.

**Review by bgeltz — CHANGES_REQUESTED** [review]

File: service/docs/source/geopm_pio_levelzero.7.rst [hunk 7]
```diff
@@ -49,6 +49,22 @@ Signals
     *  **Format**: double
     *  **Unit**: hertz
 
+``LEVELZERO::GPU_CORE_TEMPERATURE_MAXIMUM``
+    The maximum measured temperature across all sensors in the GPU accelerator."
+
+    *  **Aggregation**: expect_same
```
Comment (lhlawson) [inline service/docs/source/geopm_pio_levelzero.7.rst]:
```suggestion
    *  **Aggregation**: max
```

File: service/docs/source/geopm_pio_levelzero.7.rst [hunk 8]
```diff
@@ -49,6 +49,22 @@ Signals
...
+    *  **Domain**: gpu_chip
+    *  **Format**: double
+    *  **Unit**: celsius
+
+``LEVELZERO::GPU_MEMORY_TEMPERATURE_MAXIMUM``
+    The maximum measured temperature across all sensors in the GPU memory."
+
+    *  **Aggregation**: expect_same
```
Comment (lhlawson) [inline service/docs/source/geopm_pio_levelzero.7.rst]:
```suggestion
    *  **Aggregation**: max
```

**Review by bgeltz — APPROVED** [review]

**Comment by bgeltz** [comment]
LGTM, but I did notice that we're not consistent about initializing the ze_result_t variables.  There's one occurance of this in the constructor:
```
ze_result_t ze_result;
//Initialize
ze_result = zeInit(0);
```
But in all the other spots I saw, we declare ze_result_t's without initialization.  We might want to open an issue (or a bug depending on how you want to frame it) about properly initializing all of these.

**Comment by lhlawson** [comment]
> LGTM, but I did notice that we're not consistent about initializing the ze_result_t variables. There's one occurance of this in the constructor:
> 
> ```
> ze_result_t ze_result;
> //Initialize
> ze_result = zeInit(0);
> ```
> 
> But in all the other spots I saw, we declare ze_result_t's without initialization. We might want to open an issue (or a bug depending on how you want to frame it) about properly initializing all of these.

https://github.com/geopm/geopm/issues/2757 filed to track the issue (if we're not tackling it in this PR)

**Pull request merged**